### PR TITLE
Added some event and window attributes

### PR DESCRIPTION
**Pull request by joonis**
Window: pageXOffset, pageYOffset, scrollX, scrollY, innerWidth, innerHeight
Event: which, charCode, buttons, button (as far as practicable)

### CODE REVIEW

**Comment by ljharb** [comment]
(Closes #36)

**Comment by WebReflection** [comment]
the following
```js
this._pageX = this.clientX + window.scrollX - (html.clientLeft || 0)
```

is diffrent from
```js
this[clientXY] + (
  html[scrollLR] ||
  (document.body && document.body[scrollLR]) ||
  0
) - html[clientLR]
```

+ there are no new tests.

Mind expanding a bit on the change and tell me if you need any help adding some test?

**Comment by joonis** [comment]
I will have a look at the tests.

The following expression is assigned to `window.scrollX` through the getter function:

``` javascript
window.scrollX = html['scrollLeft'] || (document.body && document.body['scrollLeft']) || 0
```

This results in

``` javascript
this._pageX = this.clientX + (html['scrollLeft'] || (document.body && document.body['scrollLeft']) || 0) - (html.clientLeft || 0)
```

Which can be reformatted to

``` javascript
this._pageX = this.clientX + (
  html['scrollLeft'] || 
  (document.body && document.body['scrollLeft']) || 
  0
) - (html.clientLeft || 0)
```

**Comment by WebReflection** [comment]
fair enough. Please have a look if you can add some test.

To build stuff and test you need to `make dependencies` then `make` and `make test`

If you're on windows I can try to have a look myself and implement some.

**Comment by joonis** [comment]
Well, currently I don't have the time to get deeper into the test scenario. But I found a [snippet](https://github.com/WebReflection/ie8/pull/33/commits/874272d73958dbfd90340d8376facd00fbb70e91#diff-241feec3f8c247fdc432d54d90c403dbR504) from another pull request.

``` javascript
  {
    name: 'pageX, pageY and button propertyValues',
    test: function () {
      var a = document.createElement('a');
      a.className = 'target';
      a.href = 'javascript:(function(){window.Clicked=true}());';
      a.innerHTML = 'please left click here to go on with the test';
      a.addEventListener('mousedown', wru.async(function(e){
        wru.assert(!isNaN(e.pageX));
        wru.assert(!isNaN(e.pageY));
        wru.assert(e.buttons === 1);
        wru.assert(e.button === 0);
        wru.assert(e.which === 1);
        setTimeout(wru.async(function(){
          wru.assert(true);
          a.parentNode.removeChild(a);
        }), 100);
      }));
      document.body.appendChild(a);
    }
  },
  {
    name: 'charCode and which propertyValues',
    test: function () {
      var input = document.createElement('input');
      input.className = 'target';
      input.value = 'Please enter a lower "a" to continue: ';
      input.addEventListener('keypress', wru.async(function(e){
        wru.assert(e.which === 97);
        wru.assert(e.charCode === 97);
        wru.assert(typeof e.button == 'undefined');
        wru.assert(e.buttons === 0);
        setTimeout(wru.async(function(){
          wru.assert(true);
          input.parentNode.removeChild(input);
        }), 100);
      }));
      document.body.appendChild(input);
    }
  }
```

Is that the way to go?

**Comment by WebReflection** [comment]
that might work, thanks. Will have a look and double check ASAP

**Comment by joonis** [comment]
Actually the properties `pageX` and `pageY` should be calculated immediately after an event was triggered and not just on the first property access. Because the scroll position (on what these properties depend) could has been changed in the meantime. But it's a minor issue, I think.

**Comment by WebReflection** [comment]
> the scroll position (on what these properties depend) could has been changed in the meantime

nobody in years had issue with current behavior. Events expire and because of thqt are usually consumed as soon as they are triggered.

IE8 is already slow, asking it to reflow and repaint every time a mouse even happens would kill performance so it's a won't fix.

TL;DR already thought/discussed, and benchmarked, long long time ago.

**Comment by joonis** [comment]
Sure. As I wrote: minor issue.

**Comment by joonis** [comment]
And here is a test for the window properties.

``` javascript
  {
    name: 'window propertyValues',
    test: function () {
        wru.assert(window.pageXOffset === window.scrollLeft);
        wru.assert(window.pageYOffset === window.scrollTop);
        wru.assert(window.scrollX === window.scrollLeft);
        wru.assert(window.scrollY === window.scrollTop);
        wru.assert(window.innerWidth === window.clientWidth);
        wru.assert(window.innerHeight === window.clientHeight);
    }
  }
```

**Comment by WebReflection** [comment]
had to change a bit some test but it's in as `0.5.0`
